### PR TITLE
SALTO-4322: Data Records Circular Dependencies Workaround

### PR DESCRIPTION
**Pull request by tamtamirr**
Added circular dependencies workaround for additions of Records with references to Records of the same type.

---

The implementation is quite tricky, so I'll explain it in TLDR here:
1. Changed the default resolved value to be the Element itself, incase the resolved value is undefined (Which was the case for Data Instances that were added as part of this deployment)
2. Each operation now returns optionally **instancesToUpdate**, which is a list of instances that needs a further update after the initial operation.
3. The second time, we use **bulk update** to update the data instances that references to other data instances of the same type, to avoid the circular dependency.

---
_Release Notes_: 
Salesforce Adapter:
- Support deploying additions of Data Instances that reference the same type.

---
_User Notifications_: 
_None_

### CODE REVIEW

File: packages/salesforce-adapter/src/transformers/reference_mapping.ts [hunk 1]
```diff
@@ -819,7 +819,7 @@ const getLookUpNameImpl = (defs = fieldNameToTypeMappingDefs): GetLookupNameFunc
       }
       if (isElement(ref.value)) {
         const defaultStrategy = ReferenceSerializationStrategyLookup.absoluteApiName
-        return defaultStrategy.serialize({ ref, element })
+        return await defaultStrategy.serialize({ ref, element }) ?? ref.value
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/transformers/reference_mapping.ts:822]: Why this change?

File: packages/salesforce-adapter/src/transformers/transformer.ts [hunk 2]
```diff
@@ -941,31 +941,52 @@ const transformCompoundValues = async (
...
-): Promise<SalesforceRecord> => {
+): Promise<{record: SalesforceRecord; needsUpdate: boolean}> => {
+  let needsUpdate = false
   const instanceType = await instance.getType()
   const valsWithNulls = {
     ..._.mapValues(instanceType.fields, () => null),
-    ...instance.value,
+    ..._.mapValues(instance.value, val => {
+      if (isInstanceElement(val)) {
+        const referencedRecordId = val.value[CUSTOM_OBJECT_ID_FIELD]
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/transformers/transformer.ts:951]: Are references dereferenced at this stage? Meaning shouldn't the referenced value just be some random string here?
Can you explain this if a bit, I'm not sure I follow.

**Review by amitgoldarb — COMMENTED** [review]
Looks good!
Added some comments/questions.

File: packages/salesforce-adapter/src/group_changes.ts [hunk 3]
```diff
@@ -31,7 +31,7 @@ const instanceOfCustomObjectChangeToGroupId: ChangeIdFunction = async change =>
   groupId: `${change.action}_${await apiName(await (getChangeData(change) as InstanceElement).getType())}_instances`,
   // CustomObjects instances might have references to instances of the same type (via Lookup
   // fields), and if we deploy them together the reference gets removed.
-  disjoint: isAdditionChange(change),
+  disjoint: false,
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/group_changes.ts:34]: I've had a F2F discussion with @ori-moisis . Generally, (and the like the comment says) The disjoint `true` was added for the use-case we tackle in this PR. I think it'd be better to even revert @yuribro code for this file, to avoid complexity.

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts [hunk 4]
```diff
@@ -229,54 +232,49 @@ export const retryFlow = async (
...
-    return []
+    return { results: [], instancesToUpdate: [] }
   }
-  const results = await client.bulkLoadOperation(
-    typeName,
-    'update',
-    await instancesToUpdateRecords(instances)
-  )
-  return groupInstancesAndResultsByIndex(results, instances)
+  const { records, instancesToUpdate } = await instancesToRecords(instances, FIELD_ANNOTATIONS.UPDATEABLE)
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/custom_object_instances_deploy.ts:275]: Not sure it's a possible scenario considering it's existing instances, but there's no reason to explicitly ignore it IMO, to avoid unknown scenarios we are not aware of.

File: packages/salesforce-adapter/src/transformers/reference_mapping.ts (hunk 1 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/transformers/reference_mapping.ts:822]: Discussed F2F.
TLDR: In case the resolved value is undefined (Like in that case, Addition of **Data Record Instance**, we return the instance itself so it serves as a place-holder for the API name later.

File: packages/salesforce-adapter/src/transformers/transformer.ts (hunk 2 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/transformers/transformer.ts:951]: See https://github.com/salto-io/salto/pull/4491/files#r1234917246

File: packages/salesforce-adapter/src/transformers/transformer.ts [hunk 5]
```diff
@@ -941,31 +941,52 @@ const transformCompoundValues = async (
...
+      }
+      return val
+    }),
   }
   const filteredRecordValues = {
     [CUSTOM_OBJECT_ID_FIELD]: instance.value[CUSTOM_OBJECT_ID_FIELD],
     ..._.pickBy(
       valsWithNulls,
-      (_v, k) => (instanceType).fields[k]?.annotations[fieldAnnotationToFilterBy]
+      (v, k) => !isInstanceElement(v) && instanceType.fields[k]?.annotations[fieldAnnotationToFilterBy]
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/transformers/transformer.ts:965]: https://github.com/salto-io/salto/pull/4491/files#r1234917246

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts [hunk 6]
```diff
@@ -186,7 +187,8 @@ type CrudFnArgs = {
   client: SalesforceClient
 }
 
-export type CrudFn = (fnArgs: CrudFnArgs) => Promise<InstanceAndResult[]>
+export type CrudFn = (fnArgs: CrudFnArgs)
+    => Promise<{results: InstanceAndResult[]; instancesToUpdate: InstanceElement[]}>
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/custom_object_instances_deploy.ts:191]: **CrudFn** is supposed to be generic. **deployedInstancesResults** is specific for deploy operation only.

[the remark this answers is not in the document]
Comment (amitgoldarb) [reply packages/salesforce-adapter/src/custom_object_instances_deploy.ts:191]: Isn't CrudFn only in the context of deployed elements?

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (amitgoldarb) [reply packages/salesforce-adapter/src/custom_object_instances_deploy.ts:275]: This is creating a third call to the crudFn which I think we don't need. If we can think of no examples, I vote for removing this.

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts [hunk 7]
```diff
@@ -335,6 +333,13 @@ const deployAddInstances = async (
     const recordValuesWithoutNulls = cloneWithoutNulls(recordValues)
     return computeSaltoIdHash(recordValuesWithoutNulls)
   }
+  // Replacing self-reference field values with the resolved instances that will later contain the Record Ids
+  const instanceByElemId = _.keyBy(instances, instance => instance.elemID.getFullName())
+  await awu(instances).forEach(instance => {
+    instance.value = _.mapValues(instance.value, val => (isInstanceElement(val)
+      ? instanceByElemId[val.elemID.getFullName()] ?? val
+      : val))
+  })
```
Comment (tomsellek) [inline packages/salesforce-adapter/src/custom_object_instances_deploy.ts:342]: Shouldn't this be recursive? Can there not be a case where the references are inside a list/map?

File: packages/salesforce-adapter/src/transformers/transformer.ts [hunk 8]
```diff
@@ -941,31 +941,52 @@ const transformCompoundValues = async (
...
   instance: InstanceElement,
   fieldAnnotationToFilterBy: string,
-): Promise<SalesforceRecord> => {
+): Promise<{record: SalesforceRecord; needsUpdate: boolean}> => {
+  let needsUpdate = false
   const instanceType = await instance.getType()
   const valsWithNulls = {
     ..._.mapValues(instanceType.fields, () => null),
-    ...instance.value,
+    ..._.mapValues(instance.value, val => {
```
Comment (tomsellek) [inline packages/salesforce-adapter/src/transformers/transformer.ts:949]: Does this need to be recursive?

File: packages/salesforce-adapter/test/custom_object_instances.crud.test.ts [hunk 9]
```diff
@@ -513,56 +518,149 @@ describe('Custom Object Instances CRUD', () => {
...
+              expect(connection.bulk.load).toHaveBeenCalledWith(
+                'TestType__c', 'insert', expect.anything(), [
+                  { Id: undefined, Name: 'firstInstance', Number__c: 1, TestType__c: null },
+                  { Id: undefined, Name: 'secondInstance', Number__c: 1, TestType__c: null },
+                  { Id: undefined, Name: 'instanceWithoutRef', Number__c: 1, TestType__c: null },
+                ]
+              )
+              expect(connection.bulk.load).toHaveBeenCalledWith(
+                'TestType__c', 'update', expect.anything(), [
+                  { Id: 'newId0', Name: 'firstInstance', Number__c: 1, TestType__c: 'newId1' },
```
Comment (tomsellek) [inline packages/salesforce-adapter/test/custom_object_instances.crud.test.ts:659]: Silly question, but where did the ref from `newId0` to `newId1` come from?

**Review by tomsellek — COMMENTED** [review]
Had a couple of questions

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts (hunk 7 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/custom_object_instances_deploy.ts:342]: No, Salesforce Records are plain, so recursive support is redundant. List of Lookups is not a valid value for Salesforce Records as-well

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/custom_object_instances_deploy.ts:191]: v

File: packages/salesforce-adapter/test/custom_object_instances.crud.test.ts (hunk 9 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/test/custom_object_instances.crud.test.ts:659]: From the mock :)

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts [hunk 10]
```diff
@@ -360,15 +366,29 @@ const deployAddInstances = async (
...
-  const allSuccessInstances = [...successInsertInstances, ...successUpdateInstances]
+  const secondUpdateInstances = insertInstancesToUpdate.concat(updateInstancesToUpdate)
+  const {
+    successInstances: successSecondUpdateInstances,
+    errorInstances: errorSecondUpdateInstances,
+  } = await retryFlow(
+    updateInstances,
+    { typeName: await apiName(type), instances: secondUpdateInstances, client },
+    client.dataRetry.maxAttempts
+  )
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/custom_object_instances_deploy.ts:388]: @amitgoldarb After taking a deeper look into the flow, this is actually required. I'll try to explain in TLDR but we can also take this f2f :)
- In salto, the Data Record instance is added, but when querying the service, we notice it already exists (By formatting it's Elem ID from the returned Record).
- The Record can reference another instance from the same group, and can be referenced by others. We only set it's Internal Id [here](https://github.com/salto-io/salto/blob/6048650304de2920ed60dbd388d82ef4d429a0c2/packages/salesforce-adapter/src/custom_object_instances_deploy.ts#L366). The instance/s that reference this Record will have to be updated with it's internalId in another iteration.

cc. @ori-moisis please validate me here 🙏

File: packages/salesforce-adapter/src/transformers/transformer.ts [hunk 11]
```diff
@@ -946,7 +946,16 @@ const toRecord = async (
   const instanceType = await instance.getType()
   const values = {
     ...withNulls ? _.mapValues(instanceType.fields, () => null) : {},
-    ...instance.value,
+    ..._.mapValues(instance.value, val => {
+      if (isInstanceElement(val)) {
+        const referencedRecordId = val.value[CUSTOM_OBJECT_ID_FIELD]
+        if (referencedRecordId === undefined) {
+          return null
+        }
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/transformers/transformer.ts:954]: This seems like a change in the current behaviour, meaning in this "wierd" case we are sending null to the API instead of something else. Can we better explain when we will enter this if?

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts [hunk 12]
```diff
@@ -23,7 +23,7 @@ import {
   TypeElement,
   SaltoElementError,
   SaltoError,
-  SeverityLevel,
+  SeverityLevel, isInstanceElement,
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/custom_object_instances_deploy.ts:26]: Style comment - this is more consistent with current import mechanism
```suggestion
  SeverityLevel,
  isInstanceElement,
```

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts [hunk 13]
```diff
@@ -319,6 +319,15 @@ const deployAddInstances = async (
...
 ): Promise<DeployResult> => {
+  const instancesToUpdate = instances
+    .filter(instance => Object.values(instance.value).some(isInstanceElement))
+  // Replacing self-reference field values with the resolved instances that will later contain the Record Ids
+  const instanceByElemId = _.keyBy(instances, instance => instance.elemID.getFullName())
+  await awu(instances).forEach(instance => {
+    instance.value = _.mapValues(instance.value, val => (isInstanceElement(val)
+      ? instanceByElemId[val.elemID.getFullName()] ?? val
+      : val))
+  })
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/custom_object_instances_deploy.ts:330]: Reading this one again, I'm a bit confused by reading this, if the value is an InstanceElement, what are we replacing it with? What is the difference between the value we have  before this mapValue and after this mapValue?
These will later contain the recordId's because this is mutable?

**Review by amitgoldarb — COMMENTED** [review]
Looking good. A few small comments just for my understanding

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts (hunk 13 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/custom_object_instances_deploy.ts:330]: Yes. We are setting the **resolvedInstances** here, instead of the unresolved instances so they later contain the internal Ids

File: packages/salesforce-adapter/src/transformers/transformer.ts (hunk 11 above)
Comment (amitgoldarb) [reply packages/salesforce-adapter/src/transformers/transformer.ts:955]: Nvm, I understand this now.

**Review by amitgoldarb — APPROVED** [review]
I think this looks good. Let's wait for another approval though before merging this.

**Review by benisraeladi — APPROVED** [review]

File: packages/salesforce-adapter/src/group_changes.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/group_changes.ts:34]: I actually prefer leaving it like this, to be explicit
